File: apps/api/app/schemas/project.py
from datetime import datetime
from uuid import UUID
from typing import Optional
from pydantic import BaseModel, ConfigDict, Field

from app.schemas.common import PaginationParams, PaginatedResponse


class ProjectBase(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    description: Optional[str] = Field(default=None, max_length=2000)


class ProjectCreate(ProjectBase):
    pass


class ProjectUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=255)
    description: Optional[str] = Field(default=None, max_length=2000)
    status: Optional[str] = Field(default=None, pattern="^(active|archived|paused)$")


class ProjectResponse(ProjectBase):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    owner_id: UUID
    status: str
    created_at: datetime
    updated_at: datetime


class ProjectWithStats(ProjectResponse):
    target_count: int = 0
    episode_count: int = 0
    latest_episode_at: Optional[datetime] = None


class ProjectListParams(PaginationParams):
    status: Optional[str] = Field(default=None, pattern="^(active|archived|paused)$")
    search: Optional[str] = None


class ProjectListResponse(PaginatedResponse[ProjectResponse]):
    pass

File: packages/shared/src/schemas/project.ts
import { z } from "zod";
import { UUIDSchema, DateTimeSchema, PaginationParamsSchema, PaginatedResponseSchema } from "./common";

export const ProjectStatusSchema = z.enum(["active", "archived", "paused"]);
export type ProjectStatus = z.infer<typeof ProjectStatusSchema>;

export const ProjectSchema = z.object({
  id: UUIDSchema,
  name: z.string().min(1).max(255),
  description: z.string().max(2000).optional(),
  ownerId: UUIDSchema,
  status: ProjectStatusSchema.default("active"),
  createdAt: DateTimeSchema,
  updatedAt: DateTimeSchema
});
export type Project = z.infer<typeof ProjectSchema>;

export const CreateProjectRequestSchema = z.object({
  name: z.string().min(1).max(255),
  description: z.string().max(2000).optional()
});
export type CreateProjectRequest = z.infer<typeof CreateProjectRequestSchema>;

export const UpdateProjectRequestSchema = z.object({
  name: z.string().min(1).max(255).optional(),
  description: z.string().max(2000).optional(),
  status: ProjectStatusSchema.optional()
}).refine(data => Object.keys(data).length > 0, {
  message: "At least one field must be provided"
});
export type UpdateProjectRequest = z.infer<typeof UpdateProjectRequestSchema>;

export const ProjectListParamsSchema = PaginationParamsSchema.extend({
  status: ProjectStatusSchema.optional(),
  search: z.string().optional()
});
export type ProjectListParams = z.infer<typeof ProjectListParamsSchema>;

export const ProjectListResponseSchema = PaginatedResponseSchema(ProjectSchema);
export type ProjectListResponse = z.infer<typeof ProjectListResponseSchema>;

export const ProjectWithStatsSchema = ProjectSchema.extend({
  targetCount: z.number().int().nonnegative(),
  episodeCount: z.number().int().nonnegative(),
  latestEpisodeAt: DateTimeSchema.optional()
});
export type ProjectWithStats = z.infer<typeof ProjectWithStatsSchema>;